include/range/v3/view/transform.hpp

```diff
@@ -287,13 +288,17 @@ namespace ranges
                 SizedRange<Rng1 const>() && SizedRange<Rng2 const>())
             constexpr size_type_ size() const
             {
-                return std::min<size_type_>(ranges::size(rng1_), ranges::size(rng2_));
+                return ranges::on(ranges::min,
+                    ranges::compose(convert_to<size_type_>{}, ranges::size))(
+                        rng1_, rng2_);
```

Comment:
> It's a shame that we can't have both ranges::min(x, y) and ranges::min<T>(x,
y) be valid syntax when ranges::min is an object

Indeed, having to write ranges::min.operator()<T>(x, y) is just painful. It
would also be nice if operator() could be static ... I never understood
this restriction.

On Sunday, 12 June 2016, Casey Carter <notifications@github.com> wrote:

> In include/range/v3/view/transform.hpp
> <https://github.com/ericniebler/range-v3/pull/401#discussion_r66725263>:
>
> > @@ -287,13 +288,17 @@ namespace ranges
> >                  SizedRange<Rng1 const>() && SizedRange<Rng2 const>())
> >              constexpr size_type_ size() const
> >              {
> > -                return std::min<size_type_>(ranges::size(rng1_), ranges::size(rng2_));
> > +                return ranges::on(ranges::min,
> > +                    ranges::compose(convert_to<size_type_>{}, ranges::size))(
> > +                        rng1_, rng2_);
>
> Would you prefer:
>
> auto f = compose(convert_to<size_type_>{}, ranges::size);return ranges::min(f(rng1_), f(rng2_));
>
> or
>
> return ranges::min(
>     static_cast<size_type_>(ranges::size(rng1_)),
>     static_cast<size_type_>(ranges::size(rng2_)));
>
> It's a shame that we can't have both ranges::min(x, y) and ranges::min<T>(x,
> y) be valid syntax when ranges::min is an object. I wonder if anyone has
> poked EWG about making:
>
> template <class> class some_class_template;template<class T = int>
> some_class_template<T> foo{};
> auto* x = &foo;
>
> well-formed.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ericniebler/range-v3/pull/401/files/589abba031ec966799be72835b51b842fbabf118#r66725263>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AA3Npg0mcC0Uuyti4Pl8AGT71MWRdn6Eks5qLFE8gaJpZM4Iz1ib>
> .
>